### DISCUSSION TITLE
Using Factory method instead of Constuctor

### DISCUSSION THREAD

**Discussion by mehdihadeli** [kgrzybek/modular-monolith-with-ddd#214] [Q&A] [answered]
Hi @kgrzybek,
In your code base for initializing an aggregate like [meeting](https://github.com/kgrzybek/modular-monolith-with-ddd/blob/master/src/Modules/Meetings/Domain/Meetings/Meeting.cs) you have a public factory method and a private constructor in this class that factory method call this private constructor.
Why we use this factory method here instead of direct using constructor? 
What is the advantage of factory method here?

**Comment by AndreiGanichev** [comment, marked as answer]
Hi, @mehdihadeli.

@kgrzybek mentioned about some advantages [here](https://twitter.com/kamgrzybek/status/1331580335939842048?s=20):
1. You can put in Factory method some validation logic which is important to create an object from Application. Infrastructure on the contrary uses private constructors to be able restore persisted object even after validation rules were changed.
2. Factory method has a name so your code becomes more clear.

Factory method has other advantages(can be async, can return objects from cache) but for Domain model I think we can apply this one: it can hide inheritance hierarchy so you can return a subtype from the method. 
Also pay attention that often Application doesn't instantiate aggregates but delegate this to Domain model itself. Look at [CreateMeetingHandler](https://github.com/kgrzybek/modular-monolith-with-ddd/blob/master/src/Modules/Meetings/Application/Meetings/CreateMeeting/CreateMeetingCommandHandler.cs). `MeetingGroup` has `CreateMeting` method which we can also consider as Factory method because it create an instance of `Meeting`, at least from Application prospective. Check also [Don't create aggregate roots article](https://udidahan.com/2009/06/29/dont-create-aggregate-roots/) by Udi Dahan.

Of corse sometimes for simple aggregates and entities Factory method doesn't have most of these advantages and we could use ctor instead. But it's much practical to decide once that Factory method should be used to create entities from Application so you could think about complex questions in domain model instead of ctor vs Factory method choice every time. To automate such decisions there is architecture tests, look at `Entity_Should_Have_Parameterless_Private_Constructor` test in [DomainTests](https://github.com/kgrzybek/modular-monolith-with-ddd/blob/master/src/Modules/Meetings/Tests/ArchTests/Domain/DomainTests.cs)

**Reply by mehdihadeli** [reply]
@AndreiGanichev Thanks for your answer.
